src/http/Message.cc

```diff
@@ -95,21 +95,12 @@ Http::Message::parse(const char *buf, const size_t sz, bool eof, Http::StatusCod
         return false;
     }
 
-    if (hdr_len > Config.maxReplyHeaderSize || (hdr_len <= 0 && sz > Config.maxReplyHeaderSize)) {
+    if (hdr_len > Config.maxReplyHeaderSize || sz > Config.maxReplyHeaderSize) {
```

Comment:
This change introduces a bug. The old code was `(A || B)`. The proposed code is `(A || C)`, which adds a wrong condition C instead of the "correct" (but impossible to reach) condition B.

Legend:

* A: [we have found the end of the header and] the header is larger than the header size limit
* B: we have not found the end of the header but input is already larger than the header size limit
* C: input is larger than the header size limit

```suggestion
    assert(hdr_len > 0); // sanityCheckStartLine() rejects unparsed headers
    if (hdr_len > Config.maxReplyHeaderSize) {
```

N.B. The suggested assertion is essential here and also requires sanityCheckStartLine() API documentation changes in http/Message.h. Nothing in the current sanityCheckStartLine() description says that the method must treat unparsed headers as insane (i.e. return false). While you are fixing that description, please also replace sanityCheckStartLine() description in src/HttpReply.h with an `/* Http::Message API */` comment. This is how we document overloaded methods in Squid.

This PR bug fix raises a second question -- does sanityCheckStartLine() or anybody else reject large unparsable input? Checking (and, if needed, fixing) that should be done before this PR goes in IMO. I did not check.

Reply:
Thank you for your suggestions and picking up the incorrect change.

I will incorporate the changes you have suggested.

> does sanityCheckStartLine() or anybody else reject large unparsable input?
The answer to this question is: no. The only check in HttpReply::sanityCheckStartLine is for `hdr_len` being too small(`hdr_len < (size_t)(protoPrefix.psize() + 4)`). However, I am not quite sure what you are suggesting in relation to rejecting large _unparsable_ input – I am confused only about what you consider "unparsable".

If you are suggesting moving the `Config.maxReplyHeaderSize` check into sanityCheckStartLine() and/or headersEnd(), then I support this as it would free up some processing time.

This would make code like Http::One::Parser::grabMimeBlock more 'correct' AFAICT, because it does the following:
```
        if (SBuf::size_type mimeHeaderBytes = headersEnd(buf_, containsObsFold)) {
[...]
        } else { // headersEnd() == 0
            if (buf_.length()+firstLineSize() >= limit) {
                debugs(33, 5, "Too large " << which);
                parseStatusCode = Http::scHeaderTooLarge;
                parsingStage_ = HTTP_PARSE_DONE;

```

and this function is literally called like this (in once instance):
`if (!grabMimeBlock("Response", Config.maxReplyHeaderSize))`

Let me know your thoughts.

Reply:
> I am confused only about what you consider "unparsable".

I should not have used that word, sorry. I should have said something like "endless". Here, all those words essentially mean that headersEnd() returns zero. That is, we cannot find the end of headers. The end of headers may be coming later or perhaps it will never come, we do not know. What we do know is that waiting longer or reading more will not change the ultimate outcome -- the header will be rejected.

As for the right solution, I suspect that sanityCheckStartLine() should be refactored into a method that calls headersEnd() internally and then checks Config.maxReplyHeaderSize (among other things). Perhaps that refactored method would become the Message::parse() itself. Finding the right solution requires more thinking/experimentation than I have the time for right now, sorry.

I would keep the low-level headersEnd() oblivious to various high-level checks/limits, at least in this PR.

Reply:
FYI: this parse() method is deprecated. It is being replaced by the Parser class logic. Trying to polish this code up is a waste of time. Please concentrate scope down to major bugs or things compilers complain about directly.